client/virtual_channel_settlement.go

```diff
@@ -0,0 +1,188 @@
+	if err != nil {
+		c.rejectProposal(responder, err.Error())
+	}
+
+	ctx, cancel := context.WithTimeout(c.Ctx(), virtualChannelFundingTimeout)
+	defer cancel()
+
+	err = c.settlementWatcher.Await(ctx, prop)
+	if err != nil {
+		c.rejectProposal(responder, err.Error())
+	}
+
+	c.acceptProposal(responder)
+}
+
+func (c *Client) validateVirtualChannelSettlementProposal(
+	parent *Channel,
+	prop *virtualChannelSettlementProposal,
+) error {
+	// Validate parameters.
+	if prop.ChannelParams.ID() != prop.FinalState.ID {
+		return errors.New("invalid parameters")
+	}
+
+	// Validate signatures.
+	for i, sig := range prop.FinalStateSignatures {
+		ok, err := channel.Verify(
+			prop.ChannelParams.Parts[i],
+			&prop.ChannelParams,
+			&prop.FinalState,
+			sig,
+		)
+		if err != nil {
+			return err
+		} else if !ok {
+			return errors.New("invalid signature")
+		}
+	}
+
+	// Validate allocation.
+
+	// Assert equal assets.
+	if err := channel.AssetsAssertEqual(parent.state().Assets, prop.FinalState.Assets); err != nil {
+		return errors.WithMessage(err, "assets do not match")
+	}
+
+	// Assert contained before and matching funds
+	subAlloc, containedBefore := parent.state().SubAlloc(prop.ChannelParams.ID())
+	if !containedBefore || !subAlloc.BalancesEqual(prop.FinalState.Sum()) {
+		return errors.New("virtual channel not allocated")
+	}
+
+	// Assert not contained after
+	_, containedAfter := prop.State.SubAlloc(prop.ChannelParams.ID())
+	if containedAfter {
+		return errors.New("virtual channel must not be de-allocated after update")
+	}
+
+	// Assert correct balances
+	virtual := transformBalances(prop.FinalState.Balances, parent.state().NumParts(), subAlloc.IndexMap)
+	correctBalances := parent.state().Balances.Add(virtual).Equal(prop.State.Balances)
+	if !correctBalances {
+		return errors.New("invalid balances")
+	}
+
+	return nil
+}
+
+func (c *Client) matchSettlementProposal(a, b interface{}) (ok bool) {
+	// Cast.
+	inputs := []interface{}{a, b}
+	props := make([]*virtualChannelSettlementProposal, len(inputs))
+	for i, x := range inputs {
+		var prop *virtualChannelSettlementProposal
+		prop, ok = x.(*virtualChannelSettlementProposal)
+		if !ok {
+			return
+		}
+		props[i] = prop
+	}
+
+	prop0 := props[0]
+
+	for _, prop := range props {
+		// Check final state.
+		ok = prop.FinalState.Equal(&prop0.FinalState) == nil
+		if !ok {
+			return
+		}
+	}
+
+	// Store settlement state and signature.
+	virtual, err := c.Channel(prop0.FinalState.ID)
+	if err != nil {
+		return false
+	}
+	if err = virtual.machine.Update(c.Ctx(), &prop0.FinalState, hubIndex); err != nil {
+		return false
+	}
+	for i, sig := range prop0.FinalStateSignatures {
+		if err = virtual.machine.AddSig(c.Ctx(), uint16(i), sig); err != nil {
+			return false
+		}
+	}
+	if err = virtual.machine.EnableFinal(c.Ctx()); err != nil {
+		return false
+	}
+
+	// Close channel.
+	err = virtual.Close()
+	if err != nil {
+		return false
+	}
+	c.channels.Delete(virtual.ID())
```

Comment:
Not sure, but don't we also want to remove the virtual channel from persistence now? Or maybe I'm missing that this is happening somewhere else already...

Reply:
we still need the state to register the channel if the update fails, right?

Reply:
Ah so the update can still fail at this stage? Then yes. Is the virtual channel discarded from persistence after the parent channel update is successful?

Reply:
Yes, because the accept handling comes afterwards.

Before we delete the channel, we need to persist the new parent channel states. This is only done during accept handling. Alternatively, we could do the accept handling here. Then we should be fine with deleting the channel state.

Reply:
I'll reference this in an issue, so that we don't forget to plug the persistence leak.

Reply:
Added a commit where the virtual channel is withdrawn and thereby removed from persistence.